include/ekuvector/ekuvector.hpp

```diff
@@ -0,0 +1,1122 @@
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::insert(const_iterator pos, size_type count,
+                                   const Type &value) {
+  const auto &elements_to_insert = count;
+  iterator new_pos;
+  if (elements_to_insert > 0) {
+    const auto pos_ordinal = empty() ? 0 : std::distance(cbegin(), pos);
+    preallocate_capacity(size() +
+                         elements_to_insert); // this can invalidate pos
+    new_pos = begin() + pos_ordinal;
+    // move everything past pos towards the end of the vector
+    {
+      const auto end_mark = new_pos - 1;
+      auto dst = end() + elements_to_insert - 1;
+      auto src = dst - elements_to_insert;
+      while (src != end_mark) {
+        allocator_.construct(dst, std::move(*src));
+        allocator_.destroy(src); // destruct residual object
+        --src;
+        --dst;
+      }
+    }
+    // copy construct the new contents in place. The places are currently empty
+    {
+      auto dst = new_pos;
+      for (size_t i = 0; i < count; ++i, ++dst) {
+        allocator_.construct(dst, value);
+      }
+    }
+    size_ += elements_to_insert;
+  }
+  return new_pos;
+}
+
+template <class Type, class Allocator>
+template <class InputIt>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::insert(
+    const_iterator pos, InputIt first,
+    typename std::enable_if<is_iterator<InputIt>::value, InputIt>::type last) {
+  const auto elements_to_insert = std::distance(first, last);
+  iterator new_pos;
+  if (elements_to_insert > 0) {
+    const auto pos_ordinal = empty() ? 0 : std::distance(cbegin(), pos);
+    preallocate_capacity(size() +
+                         elements_to_insert); // this can invalidate pos
+    new_pos = begin() + pos_ordinal;
+    // move everything past pos towards the end of the vector
+    {
+      const auto end_mark = new_pos - 1;
+      auto dst = end() + elements_to_insert - 1;
+      auto src = dst - elements_to_insert;
+      while (src != end_mark) {
+        allocator_.construct(dst, std::move(*src));
+        allocator_.destroy(src); // destruct residual object
+        --src;
+        --dst;
+      }
+    }
+    // copy construct the new contents in place. The places are currently empty
+    {
+      auto dst = new_pos;
+      for (auto it = first; it != last; ++it, ++dst) {
+        allocator_.construct(dst, *it);
+      }
+    }
+    size_ += elements_to_insert;
+  }
+  return new_pos;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::insert(const_iterator pos,
+                                   std::initializer_list<Type> ilist) {
+  return insert(pos, ilist.begin(), ilist.end());
+}
+
+template <class Type, class Allocator>
+template <class... Args>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::emplace(const_iterator pos, Args &&... args) {
```

Comment:
@glpuga same comment about efficiency, `std::swap` may not necessarily use move assignment (which may not even be available).

Reply:
The reason for using this algorithm here is the same as before, I used the simpler algorithm for the function members that I implemented first, and did not focus on optimality for the implementation.

You're right that `std::swap` introduces the need for the move-assignment operation to be available, which is not shared by the more efficient version of the algorithm that only uses move-construct.